### ISSUE TITLE
Bug using dfu-util

### ISSUE THREAD

**Issue by astrohome** [ubis/STM32duino-PlatformIO#2]
I prefer to open a separate issue as the previous discussion was rather on Linux compatibility.

So using the latest master version, I have now an error:

`
 1
 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15

#include "Arduino.h"
 
void setup() {
  // initialize digital pin 13 as an output.
  pinMode(33, OUTPUT);
}
 
void loop() {
  /* code */
  digitalWrite(33, HIGH);   // turn the LED on (HIGH is the voltage level)
  delay(1000);              // wait for a second
  digitalWrite(33, LOW);    // turn the LED off by making the voltage LOW
  delay(1000);              // wait for a second
}
 
1.6 s
platformio run --target upload
[Sun Jul 10 16:08:28 2016] Processing f103rb (upload_protocol: dfu, upload_flags: 1, board: f103rb, framework: arduino, platform: stm32duino)
--------------------------------------------------------------------------------
Looking for dependencies...
Collecting 24 libraries
BeforeUpload(["upload"], [".pioenvs/f103rb/firmware.bin"])
Looking for upload port/disk...
 

Warning! Please install `99-platformio-udev.rules` and check that your board's PID and VID are listed in the rules.

https://raw.githubusercontent.com/platformio/platformio/develop/scripts/99-platformio-udev.rules
 
Auto-detected: /dev/ttyACM0

/home/astrohome/.platformio/packages/framework-stm32duino/tools/linux/maple_upload /dev/ttyACM0 1 0x1EAF:0x0003 /home/astrohome/code/arduino/meteostation/.pioenvs/f103rb/firmware.bin
Failed to open serial device.
dfu-util 0.8
 
Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2014 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to dfu-util@lists.gnumonks.org
 
dfu-util: Invalid DFU suffix signature
dfu-util: A valid DFU suffix will be required in a future dfu-util release!!!
dfu-util: No DFU capable USB device available
scons: *** [upload] Error 74
========================== [ERROR] Took 0.51 seconds ==========================`

I have compared the output with the correct output from Arduino IDE and found some differences.
`/home/astrohome/.platformio/packages/framework-stm32duino/tools/linux/maple_upload ttyACM0 1 1EAF:0003 /home/astrohome/code/arduino/meteostation/.pioenvs/f103rb/firmware.bin`

First of all, there is no /dev before the specific serial port, and there is no 0x in USB PID.

**Comment by ubis** [comment]
Just for a quick test, can you launch your terminal and execute the following command:

**/home/astrohome/.platformio/packages/framework-stm32duino/tools/linux/maple_upload /dev/ttyACM0 1 1EAF:0003 /home/astrohome/code/arduino/meteostation/.pioenvs/f103rb/firmware.bin**

I think thats because dfu-util reads usb id's without heximal prefix.

**Comment by astrohome** [comment]
I have tried and it worked. But after I can't anymore flash it - it just says "No dfu capable device found". And there is no serial port available.

**Comment by ubis** [comment]
Did you touched your BOOT0/BOOT1 jumpers?
Also, what command **lsusb** shows when you connect board through usb?

It might be that maple bootloader was overwritten, although it shouldn't be.

**Comment by astrohome** [comment]
I do not habe BOOT0/BOOT1 jumpers - there is a ping for BOOT1 presented on the board, and the BOOT0 should be connected to the button.

lsusb does not list Maple, using dmesg I was able to get this:

[Jul10 17:47] usb 3-1: new full-speed USB device number 17 using xhci_hcd
[  +0,129026] usb 3-1: New USB device found, idVendor=1eaf, idProduct=0003
[  +0,000002] usb 3-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[  +0,000001] usb 3-1: Product: Maple 003
[  +0,000001] usb 3-1: Manufacturer: LeafLabs
[  +0,000001] usb 3-1: SerialNumber: LLM 003
[ +22,825913] usb 3-1: USB disconnect, device number 17

I suppose that the bootloader was overwritten as there is absolutely no response from the chip.

**Comment by ubis** [comment]
The thing is, that stm32duino( or maple ) bootloader is acting wierd.
I flashed it on my STM32F103RB, and I do see it's serial port. But sometimes it doesn't shows up, instead Atmel DFU device shows up. I think serial port is required only for "touching mcu" (to reset it to flashable state). 

Also, I did flash from Arduino IDE and PlatformIO a few times. I just don't get it, sometimes serial port disappears. Thats the issue.

**Comment by albertoferna** [comment]
I have an olimexino stm32 and I have the same problem. Flashing using DFU seems to go ok. However, the board does not get enumerated any more. To solve it I have to flash the original bootloader from olimex. From the arduino IDE (with stm32duino core) I can flash without problems.

**Comment by albertoferna** [comment]
Some more information. I'm running ubuntu 16.04. I have tried to use the same command than the arduino IDE uses from the command line: `./maple_upload ttyACM0 1 1EAF:0003 /tmp/buildd5f80f66cf64b059ed7ca13825ab3b98.tmp/command_sim.ino.bin`
It works without problems, as it does in the IDE. However, the same action in platformio does not.
I have check dfu-util version. I'm using 0.8, tried both the local one installed within platformio and the system wide one, same result.
I don't know much about the behaviour of these boards. However, I would say that the problem is related to the state they get into after uploading the binary. Even if there is no response or port enumeration after uploading from platformio, I can put the board in DFU mode manually (with the buttons on the board). If I'm fast enough I can upload again from the command line or from the arduino IDE and the board would be responsive again.

**Comment by albertoferna** [comment]
I've done some more testing.
There seems to be some difference with the compiler too. I have written a very small sketch and I compiled it through both platformio and Arduino IDE. If I upload the .bin from Arduino using the command line everything works fine. However, if I do the same thing with the one compiled in platformio the board does not come up.
In fact, if I overwrite firmware.bin in platformio .pioenv folder with the one compiled in the arduino IDE and I use `pio run -target upload` the system works.
May this be some consequence of a compiler flag or similar?

**Comment by ubis** [comment]
It is due some compiller flag, I guess. 
I noticed that flashing through dfu, address is different in PlatformIO and Arduino IDE. 
This address is automatically readed by dfu-util from compiled binary.

I think I used required compiler flags, but I will re-check everything again.

**Comment by albertoferna** [comment]
I've done some more testing. Compiler (arm-none-eabi-gcc) versions are slightly different. On arduino 1.6.9 they use 4.8.3 while in platformio it is 4.8.4. I have forced platformio to use the one in arduino's IDE (by using a file system link) but still no luck.

**Comment by ubis** [comment]
Yes, I did changed compiler path that Arduino IDE would use PlatformIO's compiler and this is not a compiler-related. I'm missing some of the flags, also build size is bigger with PlatformIO.

**Comment by albertoferna** [comment]
I don't experience with this toolchain. Thus, I have no idea what most of the flags mean. I have isolate those that are different from one compiling line to another.
I have no idea if this could make a difference but in some instances the board appears as stm32f103r8 (`....STM32F1/variants/generic_stm32f103r8`) while in others as stm32f103rb (which is what is selected in platformio.ini). I have no idea if this could be the cause, just trying to help find it.

**Comment by albertoferna** [comment]
ok. Halfway there.
I manage to get dfu-util working with a firmware compiled in platformio. I had to build a new specification for a different board to be able to use the maple bootloader.
I have a different problem now, with the changes I've made, platformio tries to upload using serial_upload instead of maple_upload. I guess I have to dig a littler deeper into how boards are defined. Any clues that may help me?

**Comment by ubis** [comment]
By default, it uses serial flash method. To use dfu:

- open platformio.ini (in your project folder)
- write: upload_protocol = dfu
- write: upload_flags = 1 or upload_flags = 2.

**Comment by albertoferna** [comment]
If I try to use dfu then I have problems linking. It cannot find bootloader.ld. I believe it does not exist for a maple variant.
Since I also need to change -DVECT_TAB_ADDR to point to 0x8005000, I have created a new upload protocol called maple-dfu. I have changed just the minimum needed in main.py so that it uses the same dfu code.
I'm at a loss here. If I create a project and I don't tell it to use dfu (leave all at default), I cannot upload from platformio. However, the firmware.bin that it creates is uploaded without problems using maple-upload.
If, on the other hand, I try to use upload from platformio using dfu it creates a firmware that locks the board when uploading either from platformio or on the command line with maple-upload. ???

**Comment by albertoferna** [comment]
Upps, caught a typo on the code for maple_dfu. It is working now. Thank you very much for your help.
If anybody is interested in having support for a different board I may clean up a little whatever changes I made

**Comment by ubis** [comment]
Yes, it would be nice, if you would request pull of your changes.
